File: src/main/scala/com/juanoff/scala/factory/UserTypeFactory.scala
package com.juanoff.scala.factory

import com.juanoff.scala.types.UserType

import scala.jdk.CollectionConverters.*

class UserTypeFactory {
  private val types = scala.collection.mutable.Map[String, UserType]()

  def register(typeProto: UserType): Unit = {
    types.put(typeProto.typeName(), typeProto)
  }

  def getTypeNameList: java.util.List[String] = {
    types.keys.toList.asJava
  }

  def getBuilderByName(name: String): UserType = {
    types.get(name).orNull
  }
}

File: src/main/kotlin/com/juanoff/kotlin/factory/UserTypeFactory.kt
package com.juanoff.kotlin.factory

import com.juanoff.kotlin.types.UserType
import java.util.*

class UserTypeFactory {
    private val registry = mutableMapOf<String, UserType>()

    fun register(type: UserType) {
        registry[type.typeName()] = type
    }

    fun getTypeNameList(): List<String> = Collections.unmodifiableList(registry.keys.toList())

    fun getBuilderByName(name: String): UserType? = registry[name]
}
